solr/core/src/java/org/apache/solr/search/grouping/distributed/shardresultserializer/SearchGroupsResultTransformer.java

```diff
@@ -80,50 +106,120 @@ public NamedList transform(List<Command> data) throws IOException {
+          searchGroup.groupValue = new BytesRef(groupKey);
+        }
+      }
+      searchGroup.sortValues = getSortValues(rawSearchGroup);
+      for (int i = 0; i < searchGroup.sortValues.length; i++) {
+        final String sortField = groupSort.getSort()[i].getField();
+        SchemaField field = sortField != null ? searcher.getSchema().getFieldOrNull(sortField) : null;
+        searchGroup.sortValues[i] = ShardResultTransformerUtils.unmarshalSortValue(searchGroup.sortValues[i], field);
+      }
+      return searchGroup;
+    }
+
+    @Override
+    protected Object serializeOneSearchGroup(SearchGroup<BytesRef> searchGroup, SearchGroupsFieldCommand command) {
       Object[] convertedSortValues = new Object[searchGroup.sortValues.length];
       for (int i = 0; i < searchGroup.sortValues.length; i++) {
         Object sortValue = searchGroup.sortValues[i];
         SchemaField field = command.getGroupSort().getSort()[i].getField() != null ?
             searcher.getSchema().getFieldOrNull(command.getGroupSort().getSort()[i].getField()) : null;
         convertedSortValues[i] = ShardResultTransformerUtils.marshalSortValue(sortValue, field);
       }
-      SchemaField field = searcher.getSchema().getFieldOrNull(command.getKey());
-      String groupValue = searchGroup.groupValue != null ? field.getType().indexedToReadable(searchGroup.groupValue, new CharsRefBuilder()).toString() : null;
-      result.add(groupValue, convertedSortValues);
+      return convertedSortValues;
     }
-
-    return result;
   }
 
-}
+  public static class SkipSecondStepSearchResultResultTransformer extends DefaultSearchResultResultTransformer {
+
+    private static final String TOP_DOC_SOLR_ID_KEY = "topDocSolrId";
+    private static final String TOP_DOC_SCORE_KEY = "topDocScore";
+    private static final String SORTVALUES_KEY  = "sortValues";
+
+    public SkipSecondStepSearchResultResultTransformer(SolrIndexSearcher searcher) {
+      super(searcher);
+    }
+
+    @Override
+    protected Object[] getSortValues(Object groupDocs){
+      NamedList<Object> groupInfo = (NamedList) groupDocs;
+      final ArrayList<?> sortValues = (ArrayList<?>)groupInfo.get(SORTVALUES_KEY);
+      return sortValues.toArray(new Comparable[sortValues.size()]);
+    }
+
+    @Override
+    protected SearchGroup<BytesRef> deserializeOneSearchGroup(String groupField, Sort groupSort, String groupKey, Object groupDocs){
+      SearchGroup<BytesRef> searchGroup = super.deserializeOneSearchGroup(groupField, groupSort, groupKey, groupDocs);
+      NamedList<Object> groupInfo = (NamedList) groupDocs;
+      searchGroup.topDocLuceneId = DocIdSetIterator.NO_MORE_DOCS;
+      searchGroup.topDocScore = (float) groupInfo.get(TOP_DOC_SCORE_KEY);
+      searchGroup.topDocSolrId = groupInfo.get(TOP_DOC_SOLR_ID_KEY);
+      return searchGroup;
+    }
+
+    @Override
+    protected Object serializeOneSearchGroup(SearchGroup<BytesRef> searchGroup, SearchGroupsFieldCommand command) {
```

Comment:
https://github.com/cpoerschke/lucene-solr/commit/af1a746d8a11c04a67eeaf41c26ddbbabc7b24ce has potential (untested) tweaks for this `SkipSecondStepSearchResultResultTransformer.serializeOneSearchGroup` method such as making `uniqueField` a class member to avoid per-document get-schema-then-get-unique-key-field calls and replacement of the `DocumentStoredFieldVisitor`-based document retrieval with the simpler equivalent of [TopGroupsResultTransformer.retrieveDocument](https://github.com/apache/lucene-solr/blob/releases/lucene-solr/8.1.1/solr/core/src/java/org/apache/solr/search/grouping/distributed/shardresultserializer/TopGroupsResultTransformer.java#L285-L287) -- that would be equivalent, at least at a quick glance, unless perhaps the `DocumentStoredFieldVisitor` use has intended side effects such as cache warning (in which case maybe comments at point of use might be helpful for future readers of the code).

Reply:
Thanks 108faf8d28b7f6ba53359d76d889aedb33eb34a0, I think that `NamedList(5)` was due to a previous solution that was using 5 fields, I just used the empty constructor for simplicity.